### PR TITLE
multi local storage handling for kvm

### PR DESCRIPTION
**Pull request by DK101010**
### Description
Is currently a POC, if someone have additional ideas or hints.

**Motivation**
Users sometimes only have the option to use local storages, so they should have the option to use local storage like other storage types.

**Implementation**
Decided to use already existing code to prevent breaks in current behavior(adapted agent code by user to add local storage)
With this implementation it is possible to add and use local storage like other storage types. 

**Key properties to add a local storage**
Scope: HOST
Protocol: Filesytem
Path: mount path in KVM

**Condition:** 
Enable local storage for User VMs = true in zone detail settings

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Currently only tested in own environment

### CODE REVIEW

**Comment by weizhouapache** [comment]
@DK101010 

multiple local storage is supported since #6147 
can you test it ?

**Comment by DK101010** [comment]
> @DK101010
> 
> multiple local storage is supported since #6147 can you test it ?

@weizhouapache Damn, the last time I checked the PRs, I didn't see that one.... never mind

Hmm ... you implemented this via agent properties and my implementation works via UI like the other storages.
For user acceptance reasons I prefer my solution. User can use local storages like other storage types without touch agent properties. Also no restarts required.
Excuse me for insisting, but It will be hard to explain my users that they have to manipulate the agent properties instead to use the ui to add local storages but for other storages they have to use the ui. :/ 

Perhaps we find a compromise. But I'm afraid that our implementations are disjunct. Do you have an idea?

**Comment by weizhouapache** [comment]
> > @DK101010
> > multiple local storage is supported since #6147 can you test it ?
> 
> @weizhouapache Damn, the last time I checked the PRs, I didn't see that one.... never mind
> 
> Hmm ... you implemented this via agent properties and my implementation works via UI like the other storages. For user acceptance reasons I prefer my solution. User can use local storages like other storage types without touch agent properties. Also no restarts required. Excuse me for insisting, but It will be hard to explain my users that they have to manipulate the agent properties instead to use the ui to add local storages but for other storages they have to use the ui. :/
> 
> Perhaps we find a compromise. But I'm afraid that our implementations are disjunct. Do you have an idea?

@DK101010 

The reason I choose `agent properties` is, this can only be operated by administrator. they can do it by automation tools like chef, ansible, etc.

agree with you it would be nice to support it via api and UI, it will be more user-friendly. for example
(1) add host with multiple local storage pools on UI
(2) list all local storage pools of a host on UI (of course API as well)
(3) add local storage pool to a host (create libvirt storage pool, update db and agent.properties)
(4) remove local storage from a host (remove libvirt storage pool, update db and agent.properties)

1 is optional if 2,3,4 are supported. we can start by adding a host with only 1 local storage pool.

**Comment by DK101010** [comment]
@weizhouapache 

I would also prefer to ignore number 1.
For everything else, I can adapt the source code.

But I still have two questions. 
(on point 3. and 4. - create/remove libvirt storage pool )
Does it make sense that Cloudstack can create local storage? I had thought about that, but I came to the conclusion that it would be a break in logic. As far as I know (KVM, VMware) cloudstack does not create storages. CS only adds them to its own infrastructure. 

Why do the agent properties have to be adapted? 
My PR works completely without the agent properties and so far I have not noticed any errors that would require an adjustment in the agent properties.
I can add/remove local storages. Create/delete/migrate VM's.

**Comment by weizhouapache** [comment]
> @weizhouapache
> 
> I would also prefer to ignore number 1. For everything else, I can adapt the source code.
> 
> But I still have two questions. (on point 3. and 4. - create/remove libvirt storage pool ) Does it make sense that Cloudstack can create local storage? I had thought about that, but I came to the conclusion that it would be a break in logic. As far as I know (KVM, VMware) cloudstack does not create storages. CS only adds them to its own infrastructure.
> 
> Why do the agent properties have to be adapted? My PR works completely without the agent properties and so far I have not noticed any errors that would require an adjustment in the agent properties. I can add/remove local storages. Create/delete/migrate VM's.

@DK101010 
when cloudstack agent is started or restarted, it will 
(1) read the storage pools (uuid and path) in agent.properties
(2) create libvirt storage pools (not the directory itself) if not exist.

you can check `createLocalStoragePool` method in 
https://github.com/apache/cloudstack/blob/main/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L3419

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@DK101010 
when the coding is done, please mark this "Ready for review"

**Comment by DK101010** [comment]
@weizhouapache long time ago but now finished and already in use in our environment. Feel free to test it. 
How it looks:
image

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -648,6 +651,38 @@ public boolean stop() {
...
+        Host host = _hostDao.findByName(hostAddress);
+
+        if( host == null )
+            host = _hostDao.findByIp(hostAddress);
+
+        if( host == null )
+            host = _hostDao.findByPublicIp(hostAddress);
+
+        if( host == null )
+            throw new InvalidParameterValueException(String.format("host %s not found",hostAddress));
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:669]:
```suggestion
        if( host == null ) {
            host = _hostDao.findByIp(hostAddress);

          if( host == null ) {
              host = _hostDao.findByPublicIp(hostAddress);

            if( host == null ) {
                throw new InvalidParameterValueException(String.format("host %s not found",hostAddress));
                }
              }
            }
```
No need to recheck if host is null three times if its not null the first time.

**Review by JoaoJandre — COMMENTED** [review]
Overall looks good, just a few remarks

File: plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java [hunk 2]
```diff
@@ -550,7 +509,16 @@ private HypervisorType getHypervisorType(long hostId) {
...
     public boolean attachHost(DataStore store, HostScope scope, StoragePoolInfo existingInfo) {
-        dataStoreHelper.attachHost(store, scope, existingInfo);
+        DataStore dataStore = dataStoreHelper.attachHost(store, scope, existingInfo);
+        if(existingInfo.getCapacityBytes() == 0){
+            try {
+                storageMgr.connectHostToSharedPool(scope.getScopeId(), dataStore.getId());
+            } catch (StorageUnavailableException ex) {
+                java.util.logging.Logger.getLogger(CloudStackPrimaryDataStoreLifeCycleImpl.class.getName()).log(Level.SEVERE, null, ex);
+            } catch (StorageConflictException ex) {
+                java.util.logging.Logger.getLogger(CloudStackPrimaryDataStoreLifeCycleImpl.class.getName()).log(Level.SEVERE, null, ex);
```
[the remark this answers is not in the document]
Comment (DK101010) [reply plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:519]: Good question, I am guided by the previous code examples. What is the advantage of s_logger?

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:519]: what samples are those @DK101010 ?
in the way @JoaoJandre describes a locally stored reference to the right logger is used and no lookup in the logging component has to be done. This is general practice and I am curious in what context it was done in the way you did it, and whether that was ligit.

[the remark this answers is not in the document]
Comment (DK101010) [reply plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:519]: Hmmm, I was sure that I had copied and pasted the logger code. But now I can't find anything. Strange but ok for me. I will replace it with the s_logger.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@DK101010 , you got a checkstyle error:
```
10:05:41 [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.19.0.0-SNAPSHOT/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:74:8: Unused import - java.util.logging.Level. [UnusedImports]
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@weizhouapache , can you approve of this solution?

**Comment by weizhouapache** [comment]
> @weizhouapache , can you approve of this solution?

@DaanHoogland 
I am working on other stuff, I will look into it

**Comment by shwstppr** [comment]
@weizhouapache were you able check this?

**Comment by weizhouapache** [comment]
> @weizhouapache were you able check this?

@shwstppr sorry I could not find time for this.

**Comment by DaanHoogland** [comment]
@DK101010 , can you address the conflicts?

**Comment by DK101010** [comment]
> @DK101010 , can you address the conflicts?

I will take care of it

**Comment by DaanHoogland** [comment]
@JoaoJandre can you approve this now?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by JoaoJandre** [comment]
> @JoaoJandre can you approve this now? and maybe test?

@DaanHoogland sure, I'll review it again and test it.

File: plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java [hunk 3]
```diff
@@ -138,64 +134,26 @@ public DataStore initialize(Map<String, Object> dsInfos) {
...
-                    throw new InvalidParameterValueException("host or path is null, should be rbd://hostname/pool");
-                }
-            } else if (scheme.equalsIgnoreCase("gluster")) {
-                if (storageHost == null || storagePath == null || storageHost.trim().isEmpty() || storagePath.trim().isEmpty()) {
-                    throw new InvalidParameterValueException("host or path is null, should be gluster://hostname/volume");
-                }
-            }
-        } catch (URISyntaxException e) {
-            throw new InvalidParameterValueException(url + " is not a valid uri");
-        }
```
Comment (JoaoJandre) [inline plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:175]: Are all of these validations really unnecessary? `scheme`, ` host` and `hostPath` will never be null?

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 4]
```diff
@@ -2769,10 +2769,29 @@ private Pair<List<AsyncJobJoinVO>, Integer> searchForAsyncJobsInternal(ListAsync
...
     public ListResponse<StoragePoolResponse> searchForStoragePools(ListStoragePoolsCmd cmd) {
-        Pair<List<StoragePoolJoinVO>, Integer> result = searchForStoragePoolsInternal(cmd);
-        ListResponse<StoragePoolResponse> response = new ListResponse<StoragePoolResponse>();
+        Pair<List<StoragePoolJoinVO>, Integer> result = cmd.getHostId() != null ? searchForLocalStorages(cmd) : searchForStoragePoolsInternal(cmd);
+        return createStoragesPoolResponse(result);
+    }
+
+    private Pair<List<StoragePoolJoinVO>, Integer> searchForLocalStorages(ListStoragePoolsCmd cmd) {
+        long id = cmd.getHostId();
+        String scope = "HOST";
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:2778]: We could use the ScopeType enum here.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 5]
```diff
@@ -814,11 +855,16 @@ public PrimaryDataStoreInfo createPool(CreateStoragePoolCmd cmd) throws Resource
...
 
         DataStoreLifeCycle lifeCycle = storeProvider.getDataStoreLifeCycle();
         DataStore store = null;
         try {
-            store = lifeCycle.initialize(params);
+            if(params.get("scheme").toString().equals("file")){
+                store = createLocalStorage(params);
+            }else{
+                store = lifeCycle.initialize(params);
+            }
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:867]:
```suggestion
            if (params.get("scheme").toString().equals("file")) {
                store = createLocalStorage(params);
            } else {
                store = lifeCycle.initialize(params);
            }
```

File: plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java (hunk 3 above)
Comment (DK101010) [reply plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:175]: @JoaoJandre I moved it to StorageManagerImpl.extractUriParamsAsMap

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DK101010** [comment]
> @DK101010 Unfortunately I do not see an option to select the 'filesystem' protocol. Neither do I get an option to add via API. 'Enable local storage for User Instances' is enabled. Restarted the management server as well. image
> 
> image
> 
> Screenshot 2023-11-10 at 5 50 53 PM

Strange, we use it already in our environment without problems. hmm ... I will check this again next week.

**Comment by DaanHoogland** [comment]
> @DK101010 Unfortunately I do not see an option to select the 'filesystem' protocol. Neither do I get an option to add via API. 'Enable local storage for User Instances' is enabled. Restarted the management server as well. image
> 
> image
> 
> Screenshot 2023-11-10 at 5 50 53 PM



> Strange, we use it already in our environment without problems. hmm ... I will check this again next week.

@DK101010 , is there maybe a prereq procedure that needs documentation? like installing/enabling a provider and or setting global configurations.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DK101010** [comment]
@rajujith @DaanHoogland many thanks for testing. The problem was that the filesystem protocol is dependent on scope and zone. If the zone was not selected or the zone was selected before the scope, the protocol  is not updated. With the new scope event it should now also work for you

**Review by rajujith — APPROVED** [review]
LGTM. 
Steps I followed: 
Enable 'Enable local storage for User Instances' at zone.

Create Local storage: 

	1. Create a directory in one of the KVM host.
	2. Click on add primary storage in the cloudstack UI. 
	3. Fill the form selecting the scope as 'HOST" and the above host. Selected the protocol 'Filesystem'.
	4. Input the directory path from the above step 1. 

If the directory is not present the primary storage addition fails, I believe this is expected. 
Deleted the above primary storage, the primary storage is not visible anymore in the UI. The directory remains in the KVM host filesystem.

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@JoaoJandre can you approve now?

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Comment by DaanHoogland** [comment]
api-rate limit again:
Smoke tests completed. 115 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_trigger_shutdown | `Failure` | 336.69 | test_safe_shutdown.py
test_02_upgrade_kubernetes_cluster | `Failure` | 551.92 | test_kubernetes_clusters.py

I think we can call this a success

**Comment by andrijapanicsb** [comment]
Whoever is reviewing the PR, can we please try to ensure that the proper documentation is also provided for a feature? I'll add the related docs for this one. Thanks.

**Pull request merged**